nifi-nar-bundles/nifi-avro-bundle/nifi-avro-processors/src/main/java/org/apache/nifi/processors/avro/ConvertAvroToJSON.java

```diff
@@ -136,81 +167,54 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
+                                    : (useAvroJson ? toAvroJSON(schema, currRecord) : genericData.toString(currRecord).getBytes(StandardCharsets.UTF_8));
                             out.write(outputBytes);
-
                             if (useContainer && wrapSingleRecord) {
                                 out.write(']');
                             }
-                        }
-                    } else {
-                        try (final InputStream in = new BufferedInputStream(rawIn);
-                             final OutputStream out = new BufferedOutputStream(rawOut);
-                             final DataFileStream<GenericRecord> reader = new DataFileStream<>(in, new GenericDatumReader<GenericRecord>())) {
-
-                            int recordCount = 0;
-                            GenericRecord currRecord = null;
-                            if (reader.hasNext()) {
-                                currRecord = reader.next();
-                                recordCount++;
-                            }
-
-                            // Open container if desired output is an array format and there are are multiple records or
-                            // if configured to wrap single record
-                            if (reader.hasNext() && useContainer || wrapSingleRecord) {
-                                out.write('[');
-                            }
-
-                            // Determine the initial output record, inclusive if we should have an empty set of Avro records
-                            final byte[] outputBytes = (currRecord == null) ? EMPTY_JSON_OBJECT : genericData.toString(currRecord).getBytes(StandardCharsets.UTF_8);
-                            out.write(outputBytes);
-
-                            while (reader.hasNext()) {
-                                if (useContainer) {
-                                    out.write(',');
-                                } else {
-                                    out.write('\n');
+                        } else {
+                            try (DataFileStream<GenericRecord> stream = new DataFileStream<>(in, reader)) {
+                                int recordCount = 0;
+                                GenericRecord currRecord = null;
+                                if (stream.hasNext()) {
+                                    currRecord = stream.next();
+                                    recordCount++;
+                                }
+                                if (stream.hasNext() && useContainer || wrapSingleRecord) {
+                                    out.write('[');
+                                }
+                                byte[] outputBytes = (currRecord == null) ? EMPTY_JSON_OBJECT
+                                        : (useAvroJson ? toAvroJSON(stream.getSchema(), currRecord) : genericData.toString(currRecord).getBytes(StandardCharsets.UTF_8));
```

Comment:
Looks like toString() is still being used for encoding here, Ryan recommended a different encoder for normal JSON. Not sure what that entails though, @rdblue can you elaborate on the approach? At first glance it looks like you might need to subclass [JsonEncoder](https://github.com/apache/avro/blob/release-1.7.7/lang/java/avro/src/main/java/org/apache/avro/io/JsonEncoder.java) to handle unions differently.